### PR TITLE
Proof: genesis reveal + and groupkey reveal validation

### LINKED ISSUE #394: [tracking] Provide group key derivation at genesis proof
Wrapper issue to track
https://github.com/Roasbeef/bips/pull/33

### PR DESCRIPTION
**Pull request by halseth**
This PR builds on #381, validating the `GroupKeyReveal` and `GenesisReveal` at proof verification time.

Note that this PR doesn't validate whether the group key witness is valid, as this is a more significant change that will be in a follow-up PR.

Partially fixes #394 

TODO:
- [x] Requires #381 
- [ ] Unit tests

### CODE REVIEW

File: proof/verifier.go [hunk 1]
```diff
@@ -325,6 +325,32 @@ func (p *Proof) verifyGenesisReveal() error {
...
+	if groupKey == nil {
+		return ErrGroupKeyRequired
+	}
+
+	reveal := p.GroupKeyReveal
+	if reveal == nil {
+		return ErrGroupKeyRevealRequired
+	}
+
+	revealedKey, err := reveal.GroupPubKey(p.Asset.ID())
```
Comment (Roasbeef) [inline proof/verifier.go:342]: I think this method is missing from the diff?

Comment (halseth) [reply proof/verifier.go:342]: https://github.com/lightninglabs/taproot-assets/pull/448/commits/7a9a80795211934b0658b68a1351c1a1dc5e1e4e#diff-98586afaee9ef4f1d46ba2b4af7ce1fd31d28bc83e6df2117e6483532221f8a7R457

File: proof/verifier.go [hunk 2]
```diff
@@ -347,11 +386,26 @@ func (p *Proof) Verify(ctx context.Context, prev *AssetSnapshot,
 		return nil, err
 	}
 
-	// 5. If this is a genesis asset, and it also has a meta reveal, then
+	// 5. If this is a genesis asset, start by verifying the
+	// genesis reveal, which should be present for genesis assets.
+	isGenesisAsset := p.Asset.HasGenesisWitness()
```
[the remark this answers is not in the document]
Comment (halseth) [reply proof/verifier.go:423]: Yes, I have that change lined up for a follow-up PR

File: proof/verifier.go [hunk 3]
```diff
@@ -347,11 +386,26 @@ func (p *Proof) Verify(ctx context.Context, prev *AssetSnapshot,
...
+	hasGenesisReveal := p.GenesisReveal != nil
+
+	switch {
+	case !isGenesisAsset && hasGenesisReveal:
+		return nil, ErrNonGenesisAssetWithGenesisReveal
+	case isGenesisAsset && !hasGenesisReveal:
+		return nil, ErrGenesisRevealRequired
+	case isGenesisAsset:
```
[the remark this answers is not in the document]
Comment (halseth) [reply proof/verifier.go]: Yep! 3c5e735

**Comment by guggero** [comment]
Can be rebased on top of `v0-3-0-staging` now. The PR should also be based on that branch.

File: asset/asset.go [hunk 4]
```diff
@@ -462,6 +463,17 @@ type GroupKeyReveal struct {
...
 
+// GroupPubKey returns the group public key derived from the group key reveal.
+func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+	rawKey, err := g.RawKey.ToPubKey()
+	if err != nil {
+		return nil, err
+	}
+
+	internalKey := input.TweakPubKeyWithTweak(rawKey, assetID[:])
+	return txscript.ComputeTaprootOutputKey(internalKey, g.TapscriptRoot[:]), nil
```
Comment (guggero) [inline asset/asset.go:474]: nano nit: line length :D

File: proof/proof.go [hunk 5]
```diff
@@ -73,6 +73,25 @@ var (
...
+
+	// ErrGroupKeyRevealMismatch is an error returned if an asset proof for
+	// a genesis asset has a group key reveal that doesn't match the group
+	// key.
+	ErrGroupKeyRevealMismatch = errors.New("meta reveal doesn't match " +
+		"meta hash")
+
+	// ErrGroupKeyRevealRequired is an error returned if an asset proof for
+	// a genesis asset with a group key is missing a group key reveal.
+	ErrGroupKeyRevealRequired = errors.New("meta reveal required")
```
Comment (guggero) [inline proof/proof.go:90]: Should be "group key reveal required".

File: proof/proof.go [hunk 6]
```diff
@@ -73,6 +73,25 @@ var (
 	// for a genesis asset has a genesis reveal where the asset type doesn't
 	// match the proof TLV field.
 	ErrGenesisRevealTypeMismatch = errors.New("genesis reveal type mismatch")
+
+	// ErrNonGenesisAssetWithGroupKeyReveal is an error returned if an asset
+	// proof for a non-genesis asset contains a group key reveal.
+	ErrNonGenesisAssetWithGroupKeyReveal = errors.New("non genesis asset " +
```
Comment (guggero) [inline proof/proof.go:79]: s/has meta reveal/has group key reveal/

**Review by guggero — COMMENTED** [review]
Nice, just needs a few tweaks to the error messages (got me confused when adding the reveals) and a rebase, other than that looks good!

File: asset/asset.go [hunk 7]
```diff
@@ -418,8 +433,45 @@ type GroupKey struct {
...
+}
+
+// GroupPubKey returns the group public key derived from the group key reveal.
+func (g *GroupKeyReveal) GroupPubKey(assetID ID) (*btcec.PublicKey, error) {
+	rawKey, err := g.RawKey.ToPubKey()
+	if err != nil {
+		return nil, err
+	}
+
+	internalKey := input.TweakPubKeyWithTweak(rawKey, assetID[:])
```
Comment (jharveyb) [inline asset/asset.go:473]: Linking this comment on using existing tweak methods:

https://github.com/lightningnetwork/lnd/pull/7898#issuecomment-1686536180

**Review by jharveyb — COMMENTED** [review]
Looks close, no new nits!

Would be good to have new TODOs added in for these non-working PRs so that the tasks left for a follow-up PR are more explicit, especially since this branch won't have tests passing for the next few merges.

**Comment by dstadulis** [comment]
Reopened this PR in https://github.com/lightninglabs/taproot-assets/pull/470
Should close, save any objections.

**Pull request closed without merging**